### DISCUSSION TITLE
Why SNARIMAX does not work with any regression metric?

### DISCUSSION THREAD

**Discussion by TxusLopez** [online-ml/river#726] [Q&A]
Hi,

I'm trying to work with a SNARIMAX model in a "progressive_val_score" function:

`evaluate.progressive_val_score(
    dataset=stream.iter_pandas(scaled_features, Y),
    model=model,
    metric=metrics.MAE(),
    moment='moment',
    delay=dt.timedelta(minutes=horizon_minutes),
    print_every=1_000
)`

I find an error message saying that the SNARIMAX model does not work with MAE. Then, when I check the metrics that work with SNARIMAX, I find that any metric does it¿?:

`model = time_series.SNARIMAX(p=0,d=0,q=0)#riv_linear.LinearRegression()
metric = metrics.MSE()
print(metric.works_with(model))`

Can anyone help me? Thanks

**Comment by raphaelsty** [comment]
Hi @TxusLopez, 

The Snarimax model is not compatible with the `evaluate.progressive_val_score` function. You should use a training loop instead.

The Snarimax model is a bit of a special model. It has a `forecast` method rather than a `predict_one` method.  Snarimax is compatible with all the metrics as long as you update the metric with a single prediction and a single target at a time. I invite you to consult the documentation associated with the model [here](https://riverml.xyz/latest/api/time-series/SNARIMAX/).

You will notice that in the doc, we update the metric using the prediction at horizon 1 of snarimax, of course this choice depends on the problem you want to solve. 

```python
metric = metric.update(y, y_pred[0])
```

**Reply by MaxHalford** [reply]
Indeed, as Raphaël points out, support for time series models is lackluster.

We should probably create a utility class to allow measuring performance at multiple horizons. It shouldn't be too difficult :)

**Comment by TxusLopez** [comment]
The reason why I was trying to use `progressive_val_score` with SNARIMAX is that this function can manage the delay, that is the amount to wait before revealing the target associated with each observation to the model. And SNARIMAX creates and uses temporal features ( autoregressive and average part) and is able to forecast by using a specific time horizon. This combination (`progressive_val_score` + SNARIMAX) could be competitve for my online regression problem

**Reply by MaxHalford** [reply]
So are you just forecasting one step ahead? In that case I wouldn't frame it as a forecasting problem, but instead as a regression problem.

Again, the support for time series is a bit limited right now. It's just a specific problem that needs bespoke tools. It's difficult to integrate with the rest of the API. I'm actually in the process of revamping the `time_series` module to provide a much better experience. Stay tuned!

**Comment by TxusLopez** [comment]
Hi Max,
No, im forecasting (regression) n steps ahead, so i have to deal with delayed labels. I have tried with a hoeffding regression tree and ensembles. All of them are not able to successfully predict ahead (of course with my dataset). I think snarimax could improve their poor performance because dinamically creates and uses temporal features that could improve the results.

Enviado desde Yahoo Mail con Android 
 
  El jue., sept. 30, 2021 a 18:17, Max ***@***.***> escribió:   


So are you just forecasting one step ahead? In that case I wouldn't frame it as a forecasting problem, but instead as a regression problem.

Again, the support for time series is a bit limited right now. It's just a specific problem that needs bespoke tools. It's difficult to integrate with the rest of the API. I'm actually in the process of revamping the time_series module to provide a much better experience. Stay tuned!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by raphaelsty** [comment]
@TxusLopez, here is a wrapper to makes `time_series.SNARIMAX` compatible with `evaluate.progressive_val_score`:

```python
from river import base, time_series

class SnarimaxRegressor(time_series.SNARIMAX, base.Regressor):
    """Snarimax as a regressor for progressive_val_score models."""
    def __init__(
        self,
        p: int,
        d: int,
        q: int,
        m: int = 1,
        sp: int = 0,
        sd: int = 0,
        sq: int = 0,
        regressor: base.Regressor = None,
    ):
        super().__init__(
            p=p,
            d=d,
            q=q,
            m=m,
            sp=sp,
            sd=sd,
            sq=sq,
            regressor=regressor,
        )

    def predict_one(self, x):
        """Regressor predict_one method. The predicted horizon is set 1."""
        return self.forecast(horizon=1, xs=[x])[0]

    def learn_one(self, x, y):
        """Override learn_one method to avoid switching x and y."""
        return super().learn_one(y, x=x)
```

```python
    
from river import compose
from river import linear_model
from river import metrics
from river import evaluate
from river import preprocessing
from river import optim
from river import datasets

X_y = datasets.Bikes()
    
model = compose.Select('clouds', 'humidity', 'pressure', 'temperature', 'wind')
model |= preprocessing.StandardScaler()
model |= SnarimaxRegressor(p=0, d=0, q=0)

metric = metrics.MAE()

evaluate.progressive_val_score(X_y, model, metric, print_every=20_000)
```
[20,000] MAE: 5.090543
[40,000] MAE: 5.439107
[60,000] MAE: 5.403824
[80,000] MAE: 5.446211
[100,000] MAE: 5.455812
[120,000] MAE: 5.563714
[140,000] MAE: 5.641546
[160,000] MAE: 5.638358
[180,000] MAE: 5.577214
Out[1]:
MAE: 5.571218

The main problem here is that you will evaluate the SNARIMAX model as a regressor and not as a forecaster :)

**Reply by TxusLopez** [reply]
But can I use that wrapper to make forecasts with horizons>1 and using the delay property of `progressive_val_score`?

Thanks a lot for the support

**Reply by MaxHalford** [reply]
No, you can't. You'll have to wait till we add more support for time series.

**Reply by TxusLopez** [reply]
oh pitty, anyway thanks a lot for your support, river is great
